# PROJECT
##  Salary analysis: Exploration of Job Market and Compensation Trends

### Import Relevant Libraries

In [3]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np
import random
import statsmodels.api as sm

In [4]:
random.seed(4321)
dataset = pd.read_csv('Salary.csv')

#### Data Exploration

In [3]:
len(dataset)

6684

In [4]:
dataset.shape

(6684, 10)

In [5]:
# Including only string columns in a DataFrame description

dataset.describe(include=object)

Gender          Job_Title Country   Race Highest_Education
count    6684               6684    6684   6684              6684
unique      2                129       5     10                 4
top      Male  Software Engineer     USA  White  Bacehlors Degree
freq     3671                809    1356   1957              3021

In [6]:
# Including all the datatypes in the description 
dataset.describe()

Age  Education_Level  Years_of_Experience         Salary  \
count  6684.000000      6684.000000          6684.000000    6684.000000   
mean     33.610563         1.622382             8.076750  115307.175194   
std       7.595994         0.880474             6.030419   52806.810881   
min      21.000000         0.000000             0.000000     350.000000   
25%      28.000000         1.000000             3.000000   70000.000000   
50%      32.000000         1.000000             7.000000  115000.000000   
75%      38.000000         2.000000            12.000000  160000.000000   
max      62.000000         3.000000            34.000000  250000.000000   

            Senior  
count  6684.000000  
mean      0.143477  
std       0.350585  
min       0.000000  
25%       0.000000  
50%       0.000000  
75%       0.000000  
max       1.000000

In [30]:
# Finding the median salary value for analysis purposes
dataset['Salary'].median()

115000.0

In [7]:
# Collecting information about the dataset, datatype, null values columns if any

dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6684 entries, 0 to 6683
Data columns (total 10 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   Age                  6684 non-null   int64 
 1   Gender               6684 non-null   object
 2   Education_Level      6684 non-null   int64 
 3   Job_Title            6684 non-null   object
 4   Years_of_Experience  6684 non-null   int64 
 5   Salary               6684 non-null   int64 
 6   Country              6684 non-null   object
 7   Race                 6684 non-null   object
 8   Senior               6684 non-null   int64 
 9   Highest_Education    6684 non-null   object
dtypes: int64(5), object(5)
memory usage: 522.3+ KB


In [8]:
# Specify the sequence of the DF

pd.DataFrame(dataset, columns = ['Job_Title','Years_of_Experience',
                                 'Education_Level','Salary','Age','Gender','Country','Senior'])

Job_Title  Years_of_Experience  Education_Level  Salary  \
0         Software Engineer                    5                1   90000   
1              Data Analyst                    3                2   65000   
2                   Manager                   15                3  150000   
3           Sales Associate                    7                1   60000   
4                  Director                   20                2  200000   
...                     ...                  ...              ...     ...   
6679  Director of Marketing                   20                3  200000   
6680        Sales Associate                    3                0   50000   
6681      Financial Manager                    4                1   55000   
6682      Marketing Manager                   14                2  140000   
6683        Sales Executive                    1                0   35000   

      Age  Gender    Country  Senior  
0      32    Male         UK       0  
1      28  Female        USA       0  
2      45    Male     Canada       1  
3      36  Female        USA       0  
4      52    Male        USA       0  
...   ...     ...        ...     ...  
6679   49  Female         UK       0  
6680   32    Male  Australia       0  
6681   30  Female      China       0  
6682   46    Male      China       0  
6683   26  Female     Canada       0  

[6684 rows x 8 columns]

#### Data Analysis

##### Grouping on the basis of various columns to determine the frequency (count) per group

In [26]:
# Gender specific data distribution

gender = dataset.groupby('Gender')['Job_Title'].count()
gender.sort_values(ascending = False)

Gender
Male      3671
Female    3013
Name: Job_Title, dtype: int64

In [10]:
# Distribution of data based on race

race = dataset.groupby('Race')['Salary'].count()
race.sort_values(ascending = False) # Arrange in descending order

Race
White               1957
Asian               1599
Korean               457
Australian           452
Chinese              443
Black                435
African American     352
Mixed                334
Welsh                333
Hispanic             322
Name: Salary, dtype: int64

In [5]:
# Based on education Level

qualification = dataset.groupby('Highest_Education')['Salary'].count()
qualification.sort_values(ascending = False)

Highest_Education
BacehlorsDegree    3021
MastersDegree      1858
Doctoral           1369
HighSchool          436
Name: Salary, dtype: int64

In [12]:
# Based on Country

country = dataset.groupby('Country')['Job_Title'].count()
country

Country
Australia    1335
Canada       1322
China        1339
UK           1332
USA          1356
Name: Job_Title, dtype: int64

In [14]:
# Change Years of Experience from float to int:

dataset = dataset.astype({"Years_of_Experience": 'int'})

In [15]:
dataset.head() # View the data

Age  Gender  Education_Level          Job_Title  Years_of_Experience  \
0   32    Male                1  Software Engineer                    5   
1   28  Female                2       Data Analyst                    3   
2   45    Male                3            Manager                   15   
3   36  Female                1    Sales Associate                    7   
4   52    Male                2           Director                   20   

   Salary Country      Race  Senior Highest_Education  
0   90000      UK     White       0  Bacehlors Degree  
1   65000     USA  Hispanic       0    Masters Degree  
2  150000  Canada     White       1          Doctoral  
3   60000     USA  Hispanic       0  Bacehlors Degree  
4  200000     USA     Asian       0    Masters Degree

#### Appending 2 columns based on the conditions of values in existing columns

In [6]:
# Adding a column in text based on eduction_level

def Highest_Education(level):
    if level == 0:
        return 'High School'
    elif level == 1:
        return 'Bacehlors Degree'
    elif level == 2:
        return 'Masters Degree'
    else:
        return 'Doctoral'
    
dataset['Highest_Education'] = dataset['Education_Level'].map(Highest_Education)


# Adding a column in text based on race

def Racial_Background(race):
    if race == 'White' or race == 'Welsh' or race == 'Australian':
        return 'White'
    elif race == 'Asian' or race == 'Korean' or race == 'Chinese':
        return 'Asian'
    elif race == 'African American' or race == 'Black':
        return 'Black'
    elif race == 'Hispanic':
        return 'Hispanic'
    else:
        return 'Mixed'
    
dataset['Racial_Background'] = dataset['Race'].map(Racial_Background)
dataset


Age  Gender  Education_Level            Job_Title  Years_of_Experience  \
0      32    Male                1     SoftwareEngineer                    5   
1      28  Female                2          DataAnalyst                    3   
2      45    Male                3              Manager                   15   
3      36  Female                1       SalesAssociate                    7   
4      52    Male                2             Director                   20   
...   ...     ...              ...                  ...                  ...   
6679   49  Female                3  DirectorofMarketing                   20   
6680   32    Male                0       SalesAssociate                    3   
6681   30  Female                1     FinancialManager                    4   
6682   46    Male                2     MarketingManager                   14   
6683   26  Female                0       SalesExecutive                    1   

      Salary    Country        Race  Senior Highest_Education  \
0      90000         UK       White       0  Bacehlors Degree   
1      65000        USA    Hispanic       0    Masters Degree   
2     150000     Canada       White       1          Doctoral   
3      60000        USA    Hispanic       0  Bacehlors Degree   
4     200000        USA       Asian       0    Masters Degree   
...      ...        ...         ...     ...               ...   
6679  200000         UK       Mixed       0          Doctoral   
6680   50000  Australia  Australian       0       High School   
6681   55000      China     Chinese       0  Bacehlors Degree   
6682  140000      China      Korean       0    Masters Degree   
6683   35000     Canada       Black       0       High School   

     Racial_Background  
0                White  
1             Hispanic  
2                White  
3             Hispanic  
4                Asian  
...                ...  
6679             Mixed  
6680             White  
6681             Asian  
6682             Asian  
6683             Black  

[6684 rows x 11 columns]

In [17]:
# Data distribution based on the new racial background categories

racialgroup = dataset.groupby('Racial_Background')['Salary'].count()
racialgroup.sort_values(ascending = False)

Racial_Background
White               2742
Asian               2499
African American     787
Mixed                334
Hispanic             322
Name: Salary, dtype: int64

In [18]:
# Removing spaces before and inbetween the texts of some columns 

dataset['Job_Title'] = dataset['Job_Title'].str.split().agg("".join)
dataset['Race'] = dataset['Race'].str.split().agg("".join)
dataset['Race'] = dataset['Race'].str.strip()
dataset['Highest_Education'] = dataset['Highest_Education'].str.split().agg("".join)
dataset['Highest_Education'] = dataset['Highest_Education'].str.strip()
dataset

Age  Gender  Education_Level            Job_Title  Years_of_Experience  \
0      32    Male                1     SoftwareEngineer                    5   
1      28  Female                2          DataAnalyst                    3   
2      45    Male                3              Manager                   15   
3      36  Female                1       SalesAssociate                    7   
4      52    Male                2             Director                   20   
...   ...     ...              ...                  ...                  ...   
6679   49  Female                3  DirectorofMarketing                   20   
6680   32    Male                0       SalesAssociate                    3   
6681   30  Female                1     FinancialManager                    4   
6682   46    Male                2     MarketingManager                   14   
6683   26  Female                0       SalesExecutive                    1   

      Salary    Country        Race  Senior Highest_Education  \
0      90000         UK       White       0   BacehlorsDegree   
1      65000        USA    Hispanic       0     MastersDegree   
2     150000     Canada       White       1          Doctoral   
3      60000        USA    Hispanic       0   BacehlorsDegree   
4     200000        USA       Asian       0     MastersDegree   
...      ...        ...         ...     ...               ...   
6679  200000         UK       Mixed       0          Doctoral   
6680   50000  Australia  Australian       0        HighSchool   
6681   55000      China     Chinese       0   BacehlorsDegree   
6682  140000      China      Korean       0     MastersDegree   
6683   35000     Canada       Black       0        HighSchool   

     Racial_Background  
0                White  
1             Hispanic  
2                White  
3             Hispanic  
4                Asian  
...                ...  
6679             Mixed  
6680             White  
6681             Asian  
6682             Asian  
6683  African American  

[6684 rows x 11 columns]

In [19]:
# Writing the changes onto the csv file if required
 dataset.to_csv('Salary.csv', sep=',', index=False, encoding='utf-8')

# Linear Regression Analysis

## Using the 'statsmodels library' in Python to perform linear regression analysis. 

### In the provided code, fitting a simple linear regression model to predict the 'Salary' based on the predictor variable, here 'Age' in the dataset.

In [20]:
from statsmodels.formula.api import ols
model_fit = ols(formula="Salary ~ Age", data=dataset)
model_fit = model_fit.fit()

# Extract the optimal model parameter values, and assign them to a0, a1
a0 = model_fit.params['Intercept']
a1 = model_fit.params['Age']

# Print model parameter values with meaningful names, and compare to summary()
print( "When 0 years of age (salary) = ${:0.2f}".format(a0) )
print( "With every 1 year additional age (salary increase) = ${:0.2f}".format(a1) )
print( model_fit.summary() )

When 0 years of age (salary) = $-54876.15
With every 1 year additional age (salary increase) = $5063.39
                            OLS Regression Results                            
Dep. Variable:                 Salary   R-squared:                       0.530
Model:                            OLS   Adj. R-squared:                  0.530
Method:                 Least Squares   F-statistic:                     7550.
Date:                Wed, 20 Dec 2023   Prob (F-statistic):               0.00
Time:                        19:13:53   Log-Likelihood:                -79641.
No. Observations:                6684   AIC:                         1.593e+05
Df Residuals:                    6682   BIC:                         1.593e+05
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
---------------------------

####  Result: When the age is 0 years, the predicted salary is approximately -$54,876.15. This intercept is not practically meaningful in this context, as age is not applicable when someone has 0 years of experience. 0.53 R-squared values suggests that 53 percent of the variability in the salary can be explained by this linear regression model. Both the intercept and the coefficient for age have extremely low p-values (close to 0.000), indicating that these coefficients are statistically significant.

### Stats Model 'Years of experience' and 'Salary'

In [21]:
model_fit = ols(formula="Salary ~ Years_of_Experience", data=dataset)
model_fit = model_fit.fit()

# Extract the optimal model parameter values, and assign them to a0, a1
a0 = model_fit.params['Intercept']
a1 = model_fit.params['Years_of_Experience']

# Print model parameter values with meaningful names, and compare to summary()
print( "When 0 years of experience (salary) = ${:0.2f}".format(a0) )
print( "With every 1 year additional exp (salary increase) = ${:0.2f}".format(a1) )
print( model_fit.summary() )

When 0 years of experience (salary) = $57946.01
With every 1 year additional exp (salary increase) = $7102.01
                            OLS Regression Results                            
Dep. Variable:                 Salary   R-squared:                       0.658
Model:                            OLS   Adj. R-squared:                  0.658
Method:                 Least Squares   F-statistic:                 1.284e+04
Date:                Wed, 20 Dec 2023   Prob (F-statistic):               0.00
Time:                        19:13:53   Log-Likelihood:                -78585.
No. Observations:                6684   AIC:                         1.572e+05
Df Residuals:                    6682   BIC:                         1.572e+05
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                          coef    std err          t      P>|t|      [0.025      0.975]
------------

#### Salary vs. Education level

In [27]:
model_fit = ols(formula="Salary ~ Education_Level", data=dataset)
model_fit = model_fit.fit()

# Extract the optimal model parameter values, and assign them to a0, a1
a0 = model_fit.params['Intercept']
a1 = model_fit.params['Education_Level']

# Print model parameter values with meaningful names, and compare to summary()
print( "When lowest Education_Level (salary) = ${:0.2f}".format(a0) )
print( "With every 1 additional Education_Level (salary increase) = ${:0.2f}".format(a1) )
print( model_fit.summary() )

When lowest Education_Level (salary) = $52503.54
With every 1 additional Education_Level (salary increase) = $38710.76
                            OLS Regression Results                            
Dep. Variable:                 Salary   R-squared:                       0.417
Model:                            OLS   Adj. R-squared:                  0.417
Method:                 Least Squares   F-statistic:                     4771.
Date:                Thu, 21 Dec 2023   Prob (F-statistic):               0.00
Time:                        14:17:42   Log-Likelihood:                -80367.
No. Observations:                6684   AIC:                         1.607e+05
Df Residuals:                    6682   BIC:                         1.608e+05
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                      coef    std err          t      P>|t|      [0.025      0.975]
-------

In [23]:
dataset.head(1)

Age Gender  Education_Level         Job_Title  Years_of_Experience  Salary  \
0   32   Male                1  SoftwareEngineer                    5   90000   

  Country   Race  Senior Highest_Education Racial_Background  
0      UK  White       0   BacehlorsDegree             White

### Final Stat model with highest significance and affect on salary

In [7]:
from statsmodels.formula.api import ols
model_fit = ols(formula="Salary ~ Years_of_Experience + Highest_Education", data=dataset)
model_fit = model_fit.fit()

# Extract the optimal model parameter values, and assign them to a0, a1
a0 = model_fit.params['Intercept']
a1 = model_fit.params['Years_of_Experience']

# Print model parameter values with meaningful names, and compare to summary()
print( "When 0 years of experience and lowest education (salary) = ${:0.2f}".format(a0) )
print( "With every 1 year additional experience, salary increase = ${:0.2f}".format(a1))
print( model_fit.summary() )

When 0 years of experience and lowest education (salary) = $63408.56
With every 1 year additional experience, salary increase = $5846.05
                            OLS Regression Results                            
Dep. Variable:                 Salary   R-squared:                       0.705
Model:                            OLS   Adj. R-squared:                  0.705
Method:                 Least Squares   F-statistic:                     3999.
Date:                Wed, 03 Jan 2024   Prob (F-statistic):               0.00
Time:                        18:45:23   Log-Likelihood:                -78083.
No. Observations:                6684   AIC:                         1.562e+05
Df Residuals:                    6679   BIC:                         1.562e+05
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
                                          coef    std err          t     

#### Applying stats model to see the effect of only certain job titles with lowest p.values:

In [25]:
selected_job_titles = ['SoftwareEngineer', 'SoftwareDeveloper','DataAnalyst', 'AccountManager', 'ChiefDataOfficer', 
                      'ChiefTechnologyOfficer', 'MarketingDirector']

# Create dummies only for selected job titles
dummy_columns = pd.get_dummies(dataset['Job_Title'][dataset['Job_Title'].isin(selected_job_titles)], prefix='Job_Title', drop_first=True)

# Concatenate the dummy columns with the original dataset
dataset = pd.concat([dataset, dummy_columns], axis=1)

# Define the formula with multiple independent variables
formula = "Salary ~ Highest_Education + Years_of_Experience + Job_Title_SoftwareEngineer + Job_Title_SoftwareDeveloper + Job_Title_MarketingDirector"

# Fit the multiple linear regression model using the formula API
model_fit = ols(formula=formula, data=dataset).fit()

# Print the summary of the regression model
print(model_fit.summary())

                            OLS Regression Results                            
Dep. Variable:                 Salary   R-squared:                       0.635
Model:                            OLS   Adj. R-squared:                  0.633
Method:                 Least Squares   F-statistic:                     359.5
Date:                Wed, 20 Dec 2023   Prob (F-statistic):          2.20e-311
Time:                        19:13:53   Log-Likelihood:                -17063.
No. Observations:                1457   AIC:                         3.414e+04
Df Residuals:                    1449   BIC:                         3.418e+04
Df Model:                           7                                         
Covariance Type:            nonrobust                                         
                                         coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------------------
Inte